BTrees/_base.py

```diff
@@ -1494,3 +1545,44 @@ def MERGE_WEIGHT_default(self, value, weight):
 
 def MERGE_WEIGHT_numeric(self, value, weight):
     return value * weight
+
+def _fix_pickle(mod_dict, mod_name):
+    # Make the pure-Python objects pickle with the same
+    # class names and types as the C extensions by setting the appropriate
+    # _BTree_reduce_as attribute.
+    # If the C extensions are not available, we also change the
+    # __name__ attribute of the type to match the C name (otherwise
+    # we wind up with *Py in the pickles)
+    # Each module must call this as `_fix_pickle(globals(), __name__)`
+    # at the bottom.
+
+    mod_prefix = mod_name.split('.')[-1][:2] # BTrees.OOBTree -> 'OO'
+    bucket_name = mod_prefix + 'Bucket'
+    py_bucket_name = bucket_name + 'Py'
+
+    have_c_extensions = mod_dict[bucket_name] is not mod_dict[py_bucket_name]
+
+    for name in 'Bucket', 'Set', 'BTree', 'TreeSet', 'TreeIterator':
+        raw_name = mod_prefix + name
+        py_name = raw_name + 'Py'
+        try:
+            py_type = mod_dict[py_name]
+        except KeyError:
+            if name == 'TreeIterator':
+                # Optional
+                continue
+            raise
+        raw_type = mod_dict[raw_name] # Could be C or Python
+
+        py_type._BTree_reduce_as = raw_type
+        py_type._BTree_reduce_up_bound = py_type
+
+        if not have_c_extensions:
+            # Set FooPy to have the __name__ of simply Foo.
+            # We can't do this if the C extension is available,
+            # because then mod_dict[FooPy.__name__] is not FooPy
+            # and pickle refuses to save something like that.
+            # On the other hand (no C extension) this makes our
+            # Python pickle match the C version by default
+            py_type.__name__ = raw_name
+            py_type.__qualname__ = raw_name # Py 3.3+
```

Comment:
I'm sorry if the comments aren't clear. It's not possible to set these attributes unconditionally because if you do, then pickling the python classes fails: `Pickler.save_global` raises an exception when it finds out that `getattr(module, FooPy.__name__)` (where `FooPy.__name__` has been set to `Foo`) is not the same thing as `getattr(module, 'Foo')`.  This occurs when both the C and Python classes are present and you try to pickle the Python class; maybe this is a corner case, but it arises in the unit tests, and would arise in fixing a corrupted database and maybe elsewhere.